### ISSUE TITLE
--package flag does not work with installed packages

### ISSUE THREAD

**Issue by apljungquist** [python/mypy#7087]
In short, I expect the below to not fail

```bash
mypy -p mypy
# Can't find package 'mypy'
```

The way I read the documentation and related issues such as [6394](https://github.com/python/mypy/issues/6394), the `--package` option should find named package among installed packages and check it recursively. For instance, I expect mypy to be be able to type check itself. As demonstrated this Is not the case.

The `--module` option on the other hand works fine:
```bash
mypy -m mypy
# no output
```

Explicitly setting the `MYPYPATH` resolves the issue:

```bash
MYPYPATH=venv/lib/python3.7/site-packages mypy -p mypy
# lots of output
```

However, this solution does not work inside tox where mypy will complain `.tox/py37-mypypath-m/lib/python3.7/site-packages is in the MYPYPATH. Please remove it.`

It would be nice to be able to check typing on the package under development as an installed package because this would ensure that the package appears as typing enabled to third parties.

I uploaded a small package that also reproduces the problem [here](https://github.com/apljungquist/a_package)

OS: MacOS 10.14.5
Python: 3.7.2 (via pyenv)
Mypy: 0.720+dev.17b68c6b7eaa76853422544e32b1d6c5c3acc55a

**Comment by ilevkivskyi** [comment]
Hm, I would say this should work (unless I am missing something).

cc @gvanrossum @ethanhs

**Comment by gvanrossum** [comment]
Unfortunately the `-p` option doesn't use the same code as the import-following code, and currently only the latter supports searching `site-packages` (etc.) for packages with a `py.typed` marker in them (which mypy has). It looks like the code for `-m` does use the import-following code (by calling `BuildSource(None, m, None)`).

Agreed we should fix this. If you can come up with a PR that would be awesome!

**Comment by bhrutledge** [comment]
I discovered this while working on https://github.com/pypa/twine/pull/618/, and thinking that it would be nice to run mypy against the installed package. See [the comments](https://github.com/pypa/twine/pull/618/files/4a32958ef259fd494d748ea9bd9b68c7a263226e#r424334046) for more details (including some proposed tox configurations).

Poking around a bit, it seems like this is the entry point to the import-following code for `-p` and `-m`:

https://github.com/python/mypy/blob/22e504b09d55adc024ecd19604367651b2865db1/mypy/main.py#L900-L917

**Comment by apljungquist** [comment]
This appears to be working now. Big thanks to whoever fixed it.

**Comment by fitzymj** [comment]
It's not working for me!  I'm on mypy-0.812.

**Comment by apljungquist** [comment]
@fitzymj  Curious. In what sense does it not work for you?

I set up a [somewhat controlled job](https://github.com/apljungquist/a_package/actions/runs/802018385) to sanity check myself and my conclusion is still that it works. Notice how the older version fails because it does not find the package to type check whereas the new version fails because the type check fails.

**Comment by fitzymj** [comment]
Could be because I'm on Python 3.7, could be because I'm on Windows?  This shows it pretty clearly:

```
c:\>python -V
Python 3.7.4

c:\>mypy -V
mypy 0.812

c:\>mkdir mypackage

c:\>mkdir mypackage\testpkg

c:\>echo from . import test  1>mypackage\testpkg\__init__.py

c:\>echo print("test.py imported")  1>mypackage\testpkg\test.py

c:\>echo from setuptools import setup; setup(packages=["testpkg"])  1>mypackage\setup.py

c:\>pip install -e mypackage
Obtaining file:///C:/mypackage
Installing collected packages: UNKNOWN
  Found existing installation: UNKNOWN 0.0.0
    Uninstalling UNKNOWN-0.0.0:
      Successfully uninstalled UNKNOWN-0.0.0
  Running setup.py develop for UNKNOWN
Successfully installed UNKNOWN

c:\>python -c "import testpkg"
test.py imported

c:\>mypy -p testpkg
Can't find package 'testpkg'
```

**Comment by apljungquist** [comment]
Is `py.typed` included among the files installed for your package?

Look in a path like `venv/lib/python3.7/site-packages/testpkg`, changing `venv` as appropriate for your installation.

[This is the relevant part of the docs](https://mypy.readthedocs.io/en/stable/installed_packages.html#making-pep-561-compatible-packages).

**Comment by fitzymj** [comment]
That did it, thanks!  I went back to 0.782 and it did _not_ fix it, so I feel a little better.

Ditto on the thanks to whomever fixed it in 0.812!

**Comment by apljungquist** [comment]
You are welcome.
